src/abs.ott

```diff
@@ -107,12 +109,61 @@ formula :: formula_ ::=
   | G ( fn ) = sig :: M :: G_fn_eq_sig
     {{ coq (Map.find [[fn]] [[G]] = Some (ctxv_sig [[sig]])) }}
   | fresh ( x1 , ... , xn , e ) :: M :: fresh
-    {{ coq (my_fresh [[x1 ... xn]] [[e]]) }} % FIXME
+    {{ coq (my_fresh [[x1 ... xn]] [[e]]) }}
 
 embed
 {{ coq
-Definition my_subst (e5:e) (l:list (x*x)) := e5. (* FIXME *)
-Definition my_fresh (l : list x) (e5 : e) := True. (* FIXME *)
+Fixpoint get_val (k:x) (l:list(x*x)): option x :=
+  match l with
+  | nil => None
+  | (k', v) :: l' =>
+      match eq_x k k' with
+      | left _ => Some v
+      | _ => get_val k l'
+      end
+  end.
+
+Fixpoint my_subst (e5:e) (l:list (x*x)) : e :=
+  match e5 with
+  | e_t t => e_t t
+  | e_var x =>
+      match get_val x l with
+              | None => e_var x
+              | Some y => e_var y
+              end
+  | e_fn_call fn e_list =>
+      e_fn_call fn (map (fun e => my_subst e l) e_list)
+  end.
+
+(* we actually need to do some work to convince Coq my_fresh is well founded *)
+Fixpoint e_size (e0 : e) : nat :=
+  match e0 with
+  | e_t _ => 1
+  | e_var _ => 1
+  | e_fn_call _ es => 1 + list_sum (map e_size es)
+  end.
+
+Lemma size_ge_1: forall e0,
+     1 <= e_size e0.
+Proof. destruct e0;cbn;lia. Qed.
+
+Program Fixpoint my_fresh (l : list x) (e0 : e) {measure (e_size e0)} : Prop :=
```

Comment:
Fine by me to keep this for now, but best practices is to use [Equations](https://github.com/mattam82/Coq-Equations) instead of `Program`, since `Program` has some well-known technical issues with unfolding and the like. Should I do a PR to demonstrate, or do you want to give a try converting this to `Equations`?

Reply:
Good to know, thank you. I will give it a try

Reply:
Great, you probably want to use the `wf` directive (after registering the well-founded relation as an instance of `WellFounded`), like here: https://github.com/mattam82/Coq-Equations/blob/main/examples/general_recursion.v#L165